<Tooltips.Concept /> New feature for <ConceptMap />

neenjaw opened the pull request:
This PR adds tooltip/popover/dialog-type things to the <ConceptMap /> React component

This is based off of the work done for the other tooltips, but it has diverged a little for a couple reasons:
- they don't directly overlay
- they have some mild interactivity (links)
- they pull HTML content directly from the API using `react-query`
- the tooltip is self-controlled component, rather than a controlled component from the parent

---

## Goals of the feature

- Accessibility, this was designed first as a tab-navigable feature, then mouse interactivity added after.
- Consistent behavior
  - Tab-navigation takes precedence.  When the concept/tooltip has focus, the tooltip is open, mouse cannot open more tooltips u ntil focus is lost (blur)
  - Mouse navigation must allow for the tooltip to be detached
- Must allow for direct HTML injection from the API
- tooltips must adjust their placement based on the window size
  - This is done using popper, like the other tooltips

## Screenshots of features

> popper below (default)

![tooltip-show-popper-bottom](https://user-images.githubusercontent.com/8953691/99124619-2044d180-25c8-11eb-9310-e31b16ff6c35.png)

> popper above (when window position doesn't allow)

![tooltip-show-popper-top](https://user-images.githubusercontent.com/8953691/99124642-2c309380-25c8-11eb-9506-23efb7572092.png)

> Showing focus preference to mouse

![tooltip-show-focus-preference](https://user-images.githubusercontent.com/8953691/99124659-36529200-25c8-11eb-94a5-7f786dbe6f36.gif)

> Showing tab-navigation (forward only, but works in reverse too)

![tooltip-show-tab-nav](https://user-images.githubusercontent.com/8953691/99124700-45d1db00-25c8-11eb-9ffc-f4e85af1e480.gif)

> Showing hover behavior

![tooltip-show-hover](https://user-images.githubusercontent.com/8953691/99124720-4e2a1600-25c8-11eb-9a20-9f187f751b1f.gif)

---

New files introduced:

- `app/javascript/components/tooltips/Concept.tsx`
  - This is the main new component, arguably I think there is room here for some codesplitting
    - Perhaps the state reducer and open-tooltip management functions could be moved into another file

Refactoring done:

- `app/javascript/components/tooltips/index.js` :arrow_right:  `app/javascript/components/tooltips/index.ts`
  - prevent loss of type information
- `app/commands/track/determine_concept_map_layout.rb`
  - added tooltip url to concept map generation
  - tests fixed to support change
- `app/javascript/components/tooltips/MentoredStudent.tsx`
  - added exported function return type
- `<ConceptMap />` CSS moved to the `css` dir :arrow_right: `app/css/components/track/concept-map.css`
  - imports updated

neenjaw committed:
add tooltip url to concept

neenjaw committed:
fix concept map css file location

neenjaw committed:
WIP - Tooltips.Concept added

Also now drilling down tooltip endpoint from ConceptMap.

Problem: when using react-query, getting back a readable stream.
Possible that the header for the response is set to binay.

neenjaw committed:
WIP - 🤦 all I needed to do was call `res.text()`

neenjaw committed:
WIP - Concept Tooltip is having issue with setState inside of useEffect.

neenjaw committed:
It works?

iHiD commented on app/css/components/tooltips/concept.css:
```diff
@@ -5,6 +5,8 @@
 
     @apply shadow-lg rounded-8 py-20 px-24 bg-white;
 
+    transition: opacity 120ms cubic-bezier(0.4, 0, 0.2, 1);
```
Could we comment this out (but leave it there) and put a `/* TODO: Add fade in */` here pls. I'm trying to keep all design stuff to only things that that Tai has approved for now.

neenjaw replied:
Sure

SleeplessByte replied:
Thank you for using the "correct" transition tho.

iHiD commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -50,6 +57,9 @@ export const Concept = ({
     classes.push('active')
   }
+  if (isActiveHover) {
+    classes.push('hover')
```
What does this class do?

neenjaw replied:
Nothing anymore, my plan was to remove it

iHiD commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -60,44 +70,44 @@ export const Concept = ({
+// focus on concept / focus-blur
+// focus on tooltip / focus-blur
+// focus inside tooltip
+// no focus
```
Can these be removed now?

neenjaw replied:
Ya, I'll clean them up when this is ready for final review. It's only half working so far

iHiD replied:
OK :)

neenjaw committed:
change transition to comment

neenjaw committed:
🤯 pointer-events-auto for the tool tip....

neenjaw committed:
✔️ Functional tooltip/popover/dialog
mouseover works, tab-navigation works

neenjaw committed:
🎉 It is now accessible too!

neenjaw committed:
tests fixed

neenjaw commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+  const popper = usePopper(parent, tooltipRef.current, {
+    placement: 'bottom',
+    modifiers: [{ name: 'offset', options: { offset: [0, 8] } }],
```
This line changes the offset of the tooltip from the concept.

neenjaw commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+    setTimeout(
+      () => dispatch(tooltipId, showDispatcher, 'request-hover', 'hide'),
+      150
```
this is the delay once the mouse leaves the concept before tool-tip disappears (unless it has another reason to stay open). It facilitates the ability to move the mouse across the gap.  I just picked a duration that felt good. Up for change.

iHiD replied:
This is nice. Can you add this as a comment to the code please. The first two sentences of your comment here would be perfect 🙂 

Always good to add a comment for any "magic numbers" that aren't worth turning into a variable name (and I agree that here, inlining is best).

ErikSchierboom replied:
Aha, that must be what's causing this:

![tooltip](https://user-images.githubusercontent.com/135246/99370499-0bb15380-28be-11eb-9d0a-b434a2b90ca0.gif)

neenjaw replied:
Ya, should just tweak the timing there a bit

neenjaw replied:
Thanks for the gif, makes it easier to see what's happening!

ErikSchierboom replied:
First time making a GIF!

![Zvxx53q](https://user-images.githubusercontent.com/135246/99405050-990b9c80-28ec-11eb-9668-00794fb7ae4d.gif)

neenjaw replied:
This should now be fixed in https://github.com/exercism/v3-website/pull/273/commits/d803d7780035a47d9f9e87f7cd019873e43b8fd6

neenjaw commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  action?: ShowActionValue
+): void {
+  dispatcher([id, dispatcher, source, action])
+}
```
This could probably be moved to another file

SleeplessByte replied:
This entire function should probably go (it doesn't do anything except for forwarding the arguments given).

neenjaw commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  // Retrieve the HTML contents from the endpoint
+  const { isLoading, isError, data: html_content } = useQuery(tooltipId, () =>
+    fetch(endpoint).then((res) => res.text())
+  ) as { isLoading: boolean; isError: boolean; data: string }
```
If there is merit, this could be abstracted to a custom hook, like in the other tooltips

iHiD commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -1,5 +1,7 @@
 import React, { useEffect, useRef, useState, MouseEventHandler } from 'react'
 import { CompleteIcon } from './CompleteIcon'
+import * as Tooltips from '../tooltips'
```
Should this just import `Tooltips.Concept` rather than everything?

neenjaw replied:
It could be, I just followed the already set pattern.

iHiD replied:
Let's import the single thing. Could you change this when you get a min pls? :)

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  if (nextState.show) {
+    addOpenTooltip(id, dispatch)
```
This action is a side effect. There shouldn't be side-effects inside the reducer. If you want to _act_ on a _state change_, instead, use a `useEffect` or `useLayoutEffect` outside of the reducer.

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  if (nextState.show) {
+    addOpenTooltip(id, dispatch)
+  } else {
+    removeOpenTooltip(id)
```
Same

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+  // request-focus and focus actions take precedence over mouse events
+  // so it can close any other for accessibility purposes
+  if (
```
Same

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  // Retrieve the HTML contents from the endpoint
+  const { isLoading, isError, data: html_content } = useQuery(tooltipId, () =>
+    fetch(endpoint).then((res) => res.text())
+  ) as { isLoading: boolean; isError: boolean; data: string }
```
Doesn't `useQuery` return this as a type?? I don't think you need to cast it.

neenjaw replied:
It was erroring without it I thought. :shrug: I can try again. Maybe I just need to restart `TSServer` :smile:

SleeplessByte replied:
You need:

```typescript
useQuery<ReturnType>()
```

or even

```typescript
useQuery<ReturnType, ErrorType>()
```

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  hoverRequestToShow,
+  focusRequestToShow,
+  parentSlug = '',
+}: {
```
Extract to `ConceptProps`

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  ) as { isLoading: boolean; isError: boolean; data: string }
+
+  const tooltipRef = useRef(null)
+  const [{ show }, showDispatcher] = useReducer(showReducer, {
```
consider extracting the initial state to a constant or a function that returns this (reducers can take functions for intiial state)

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      role="tooltip"
+      tabIndex={show ? undefined : -1}
+      onFocus={() => {
+        dispatch(tooltipId, showDispatcher, 'focus', 'show')
```
Consider instead renaming `showDispatcher` to `dispatch` and changing the array to an object (think positional arguments to keyword arguments):

```javascript
dispatch({ type: 'show', cause: 'focus', target: tooltipId })
```

iHiD replied:
(For clarity, this is the same as my comment below but with the addition of using the object)

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      tabIndex={show ? undefined : -1}
+      onFocus={() => {
+        dispatch(tooltipId, showDispatcher, 'focus', 'show')
+      }}
```
```suggestion
      onFocus={() => { showDispatcher(tooltipId, 'focus', 'show') }}
```

neenjaw replied:
it's just how prettier formatted it for me when I hit `ctrl+s`, I'll see if I can get that to stick

SleeplessByte replied:
Don't. It's prettier. You don't need to always agree with it :)

neenjaw replied:
:relieved:

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    setTimeout(
+      () => dispatch(tooltipId, showDispatcher, 'request-hover', 'hide'),
```
Should guard against unmount before the timer ends.

iHiD replied:
```suggestion
      () => showDispatcher(tooltipId, 'request-hover', 'hide'),
```

SleeplessByte commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+    dispatch(tooltipId, showDispatcher, 'request-focus', 'hide')
+  }, [focusRequestToShow]) // eslint-disable-line react-hooks/exhaustive-deps
+
```
Should hide itself on unmount

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+        dispatch(tooltipId, showDispatcher, 'focus', 'hide')
+      }}
+      onMouseEnter={() => dispatch(tooltipId, showDispatcher, 'hover', 'show')}
+      onMouseLeave={() => dispatch(tooltipId, showDispatcher, 'hover', 'hide')}
```
As far as I can tell, `dispatch` is unneccessary and can be replaced as follows

```suggestion
      onFocus={() => { showDispatcher(tooltipId, 'focus', 'show') }}
      onBlur={() => { showDispatcher(tooltipId, 'focus', 'hide') }}
      onMouseEnter={() => showDispatcher(tooltipId, 'hover', 'show')}
      onMouseLeave={() => showDispatcher(tooltipId, 'hover', 'hide')}
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  useEffect(() => {
+    if (focusRequestToShow && popper?.update) {
+      popper.update()
+      dispatch(tooltipId, showDispatcher, 'request-focus', 'show')
```
```suggestion
      showDispatcher(tooltipId, 'request-focus', 'show')
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      return
+    }
+
+    dispatch(tooltipId, showDispatcher, 'request-focus', 'hide')
```
```suggestion
    showDispatcher(tooltipId, 'request-focus', 'hide')
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  useEffect(() => {
+    if (hoverRequestToShow && popper?.update) {
+      popper.update()
+      dispatch(tooltipId, showDispatcher, 'request-hover', 'show')
```
```suggestion
      showDispatcher(tooltipId, 'request-hover', 'show')
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+const closeOtherOpenTooltips = (id: string) => {
+  openTooltips.forEach(([openId, openDispatcher]) => {
+    if (openId !== id) {
+      dispatch(openId, openDispatcher, 'force-close')
```
```suggestion
      openDispatcher(openId, 'force-close')
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  state: ShowState,
+  [id = '', dispatch, source, action]: [
+    string,
+    React.Dispatch<unknown>,
```
```suggestion
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+ */
+function showReducer(
+  state: ShowState,
+  [id = '', dispatch, source, action]: [
```
This `dispatch` object isn't used in this method.

```suggestion
  [id = '', source, action]: [
```

iHiD replied:
Does the `id` need a default value? What is the situation where that happens?

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  ]
+) {
+  const nextState: ShowState = { ...state }
+  const parsedAction = action === 'show'
```
I found this variable name confusing.

```suggestion
  const shouldShow = action === 'show'
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+  switch (source) {
+    case 'request-hover':
+      nextState.requestHover = parsedAction && !hasOpenTooltip(id)
```
```suggestion
      nextState.requestHover = shouldShow && !hasOpenTooltip(id)
```

neenjaw replied:
Thanks! This was like the 5th name and I just couldn't find one I liked.  Hindsight 20-20 :eyeglasses:

iHiD replied:
Yeah, totally. (To continue to metaphor) a second pair of eyes is invaluable for naming!

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      nextState.show = nextShow(nextState)
+      break
+    case 'request-focus':
+      nextState.requestFocus = parsedAction
```
```suggestion
      nextState.requestFocus = shouldShow
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      nextState.show = nextShow(nextState)
+      break
+    case 'hover':
+      nextState.hover = parsedAction && !hasOpenTooltip(id)
```
```suggestion
      nextState.hover = shouldShow && !hasOpenTooltip(id)
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+      nextState.show = nextShow(nextState)
+      break
+    case 'focus':
+      nextState.focus = parsedAction
```
```suggestion
      nextState.focus = shouldShow
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  action?: ShowActionValue
+): void {
+  dispatcher([id, dispatcher, source, action])
+}
```
I can't see the purpose of this method. It seems to just indirectly call a method on one of its paramaters.

```suggestion
```

iHiD replied:
Regardless of this change, here we have `action?` but in the dispatcher, we're missing the `?`. I would expect them both to be the same, but I'm quite possibly wrong :)

neenjaw replied:
Ya, it started as something else then mutated over time, you and @SleeplessByte are right.  It just needs to go.

iHiD replied:
That's cool. Refactoring to delete code once everything works is the perfect way to solve things. Red/green/refactor FTW! :)

neenjaw replied:
removed

iHiD reviewed (commented):
OK, first pass review. This removes a bit of indirection that confused me. I've started here because the rest didn't confuse me in the same way! :)

I'll look at more of it next.

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  parent,
+  hoverRequestToShow,
+  focusRequestToShow,
+  parentSlug = '',
```
The "parent" and the "slug" are only related here via some coincidence. This tooltip might reasonable appear on things that are not concepts. So this would be better loosely coupled from that, and named `slug` (or possible `conceptSlug`)

```suggestion
  slug = '',
```

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+
+export const Concept = ({
+  endpoint,
+  parent,
```
I don't think this is actually a `parent`. It's not hierachical (the element actually sits next to it in the tree) and parent is quite non-descript as to the purpose it serves in this class. I think we can name it relevantly to *this* class. So  It's the `triggeringElement` (in the context of a tooltip) or `referenceElement` (popper's name for it) etc

neenjaw replied:
ya, `referenceElement` or `triggeringElement` is probably the way to go

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  }
+  if (show) {
+    classNames.push('tw-pointer-events-auto')
+  }
```
~Two~ Three comments here.

1. If/else

```suggestion
  if (show) {
    classNames.push('tw-pointer-events-auto')
  }
  else {
    classNames.push('tw-invisible')
  }
```

2. Do we need to turn `tw-pointer-events-auto` off? If the element has `display:none`, then I don't think it matter if these are on. So I think this can just be added to the default set on L200, and you can *only* have the `if(!show)..` bit

3. Is there a reason you're using `tw-invisible` and not `tw-hidden`? I think we should use `display:none` here unless there's a good reason not to.

4. I think this should have a default state and a pivot. I'm presuming we consider tooltips to be shown by default, so we can just put `tw-pointer-events-auto` in the stylesheet?

iHiD replied:
Also, if we have (2) on always (as opposed to requiring conditional logic), that should be added to the CSS class for `c-concept-tooltip` not added here..

neenjaw replied:
1) Sure, I think they were separated by something else which I removed
2) So, `pointer-events: none` is set on the figure, so it is inherited by the tooltip, and strangely, without it set to none seems to trigger it to show even when hidden.
3) Using **invisible** so that popper can calculate the position of the tooltip even when not seen and to make it invisible to screen-readers (rather than opacity)
4) I hear what you're saying, but I think what I'm saying is that without manipulating the pointer events (as in `(2)`) it causes some triggering problems.

iHiD replied:
> without it set to none seems to trigger it to show even when hidden.

My guess would be because it's `invisible` and not `hidden`

> so that popper can calculate the position of the tooltip even when not seen

Why is this desirable?

neenjaw replied:
This system has been changed a bit, I can't really comment on this implmentation now.  Will explain when the PR is ready again for re-review

neenjaw replied:
So now, if the tooltip is not hovered, the div is not mounted to the DOM, when it is, it is initially invisible so it can be positioned by popper (visibility: hidden) then made visible.

iHiD commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  return (
+    <div
```
Should we use `<dialog>` for these?

iHiD replied:
Apparently, we shouldn't because they don't cause a change in focus.

neenjaw replied:
I don't know. I thought dialog has a different meaning with semantic HTML tho?

iHiD commented on test/commands/track/determine_concepts_layout_test.rb:
```diff
@@ -28,7 +28,8 @@ def test_layout_with_one_level
               name: 'Basics',
-              web_url: 'https://test.exercism.io/tracks/ruby/concepts/basics'
+              web_url: 'https://test.exercism.io/tracks/ruby/concepts/basics',
+              tooltip_url: '/tracks/ruby/concepts/basics/tooltip'
```
Can we make these URLs, not paths pls? (You'll obviously need to change the code to match -`_path` to `_url`.)

neenjaw replied:
Oh, I think I just copied the ruby helper code from your example.  I didn't know that the ending changed things! :laughing:

neenjaw committed:
start of revisions:
- removed dispatch redirector
- dispatch now takes object rather than positional args
- Tooltips.Concept now uses reference____ vs parent____

neenjaw committed:
path to url change
was: Exercism::Routes.tooltip_track_concept_path
is now: Exercism::Routes.tooltip_track_concept_url

neenjaw committed:
extract function interfaces

neenjaw committed:
reducer type system added

iHiD commented on test/commands/track/determine_concepts_layout_test.rb:
```diff
@@ -57,12 +58,14 @@ def test_layout_with_two_levels
               name: 'Basics',
-              web_url: 'https://test.exercism.io/tracks/ruby/concepts/basics'
+              web_url: 'https://test.exercism.io/tracks/ruby/concepts/basics',
+              tooltip_url: '/tracks/ruby/concepts/basics/tooltip'
```
```suggestion
              tooltip_url: 'https://test.exercism.io/tracks/ruby/concepts/basics/tooltip'
```

etc

neenjaw replied:
Ya, haven't finished all the changes just yet. Ran out of time last night. Will be working on this more today

iHiD replied:
No probs! I figured you knew :)

neenjaw committed:
update import

neenjaw committed:
useReducer rewrite
This is a pretty substantial rewrite which basically amounts to the same
result as the previous version, but there are several improvements:
1) useReducer's reducer is now a pure function (no side effects)
2) it consistently use actions to set the state of the tooltip, then the
  state is acted on by useEffect hooks
3) usePopper now follows their documented pattern
4) the tooltip is only rendered to the dom in the invisible and visible
  states, it is destroyed otherwise
5) type system and type naming improved

neenjaw committed:
fix test url

ErikSchierboom commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -10,26 +12,31 @@ import {
   slug,
   name,
   web_url,
+  tooltip_url,
```
We're using camelCase for the other properties. Can we use them here too?

neenjaw replied:
I think these are like this because these properties are coming from the rails api. So either I need to use camelcase in rails, or snake case in typescript.

SleeplessByte replied:
@kntsoriano has a helper for this!

neenjaw replied:
help me @kntsoriano, you're my only hope!

ErikSchierboom replied:
@neenjaw See https://github.com/exercism/v3-website/blob/784415bd5b845e978d8ea3949fa3465600c3e9de/app/javascript/hooks/request-query.ts#L37

kntsoriano replied:
@neenjaw I use this to import the helper: `import { camelizeKeys } from 'humps'`

We usually apply the helper when we get the response from the API. But in your case, there is no API request, so I would probably camelize the keys inside `application.tsx`. Hope that helps!

neenjaw replied:
Yes! Thanks!

SleeplessByte replied:
(This).

neenjaw replied:
done

neenjaw replied:
Thanks for the tips!

ErikSchierboom commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,423 @@
+  }`
+
+  // Retrieve the HTML contents from the contentEndpoint
+  const { isLoading, isError, data: html_content } = useQuery<string>(
```
Could `html_content` use camelCase?

```suggestion
  const { isLoading, isError, data: htmlContent } = useQuery<string>(
```

neenjaw replied:
Yes, I think I just had this separate because of its "unsafe" usage in my head.

SleeplessByte replied:
(it's not unsafe tho :P)

neenjaw replied:
fixed

ErikSchierboom commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,423 @@
+  return nextState
+}
+
+export const Concept = ({
```
Just to check: most of the code is not specific to the Concept tooltip, but could be applied to other tooltips too, right? Were you planning to extract this to a separate component, for it to be re-used?

neenjaw replied:
We had discussed this in slack, and the one case that @iHiD had been thinking of we weren't sure if it just deserved just another type of tooltip. So I haven't made an active effort to ensure generalization

ErikSchierboom replied:
Makes sense!

ErikSchierboom commented on app/javascript/components/tooltips/Concept.tsx:
```diff
@@ -0,0 +1,423 @@
+  // Retrieve the HTML contents from the contentEndpoint
+  const { isLoading, isError, data: html_content } = useQuery<string>(
+    tooltipId,
+    () => fetch(contentEndpoint).then((res) => res.text())
```
The editor uses the `AbortController` class to allow cancellation of pending `fetch` requests. It might be nice to add that here too. See https://github.com/exercism/v3-website/blob/84863725436106c30e5e04bc41ca43031d1acd0d/app/javascript/components/student/Editor.tsx#L187

ErikSchierboom commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -60,41 +67,40 @@ export const Concept = ({
+        hoverRequestToShow={isActiveHover}
+        focusRequestToShow={hasFocus}
+        referenceElement={conceptRef.current}
+        referenceConceptSlug={slug}
```
Nice and simple way to include the tooltip! 👍

ErikSchierboom commented on app/javascript/components/concept-map/ConceptMap.tsx:
```diff
@@ -68,6 +68,7 @@ export const ConceptMap = ({
                     slug={slug}
                     name={concept.name}
                     web_url={concept.web_url}
+                    tooltip_url={concept.tooltip_url}
```
Could we use camelCase here?

```suggestion
                    tooltipUrl={concept.tooltip_url}
```

ErikSchierboom reviewed (approved):
Love it! Left some minor comments.

neenjaw committed:
fix casing

neenjaw committed:
hover issues (💢) fixed with pseudo elements (💃)
- previously a timeout was used to cover allow the mouse to cross over
  over the gap between the concept and the tooltip. This resulted in
  strange behavior at times due to conflicting states on hover show etc|
- now transparent pseudo element is placed in the :before :after to
  allow the mouse to cross without that same issue. 🎉

iHiD commented on app/css/components/tooltips/concept.css:
```diff
@@ -91,4 +93,27 @@
+        height: 8px;
+        left: 0;
+        top: 0;
+        display: inline-block;
```
Is this needed?

neenjaw replied:
Yes so that it can be used to cover the gap between the `<Concept>` and `<ConceptTooltip>`. Otherwise the tooltip closes before you have a chance to hover over it when there is a gap

neenjaw replied:
I think `display: block` might work too, but untested

iHiD replied:
Yeah. I think you can remove this line altogether and it will default to `display: block`.

neenjaw committed:
remove `inline-block`, revert to `initial`

SleeplessByte reviewed (commented):
Tell me when I should look at this.

SleeplessByte commented on app/css/components/tooltips/concept.css:
```diff
@@ -91,4 +93,25 @@
         }
     }
+
+    &::before {
```
I tend to comment these so the next reader understands _why_ this before and after pseudo element is necessary. In this case it's not because of visuals, so it's important!

SleeplessByte commented on app/css/components/tooltips/concept.css:
```diff
@@ -91,4 +93,25 @@
+        left: 0;
+        top: 0;
+        transform: translateY(-100%);
+        background: rgba(0, 0, 0, 0);
```
```suggestion
        background: transparent;
```

![Showing the colour picker for the transparent keyword colour. It matches #00000000, which is rbga(0, 0, 0, 0)](https://user-images.githubusercontent.com/1964376/99968089-fa25eb00-2d98-11eb-8e7a-f73fa5318858.png)

It's usually easier to parse with intent.

SleeplessByte commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -10,26 +12,31 @@ import {
 }): JSX.Element => {
   const [visibility, setVisibility] = useState<Visibility>('hidden')
   const conceptRef = useRef(null)
+  const [hasFocus, setHasFocus] = useState<boolean>(
```
This is ambiguous. Usually when an element is focusable and it contains focusable elements, we'd say "focus within", but in this case I think focus is meant to be "focus is on this concept (or within)". Its initial value is "focus is on this concept (but NOT within)".

So:
- either the logic isn't completely sound
- or the naming isn't perfect

It's fine to keep, but add a comment!

SleeplessByte commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -10,26 +12,31 @@ import {
+  isActiveHover,
   exercises = 0,
   exercisesCompleted = 0,
 }: IConcept & {
```
Perhaps `ConceptProps` is a good idea now. This list has become long.

neenjaw committed:
add css comment re: req for function

neenjaw committed:
rgba to transparent

neenjaw committed:
snake case to camel case, extract <Concept /> Prop

neenjaw committed:
add abort controller
when there is a navigation away from the concept map while the tooltips
are loading, this will cancel them
